// @smm/sa/commonjs/sa.js
(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);throw new Error("Cannot find module '"+o+"'")}var f=n[o]={exports:{}};t[o][0].call(f.exports,function(e){var n=t[o][1][e];return s(n?n:e)},f,f.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
exports.__esModule = true;
var DeviceCookieID = "SMM_device_id";
exports.DeviceCookieID = DeviceCookieID;
var TokenCookieID = "SMM_auth_token";
exports.TokenCookieID = TokenCookieID;
var SessionCookieID = "SMM_session_id";
exports.SessionCookieID = SessionCookieID;
var SessionStartTimeCookieID = "SMM_session_start_timestamp";
exports.SessionStartTimeCookieID = SessionStartTimeCookieID;
var MetalCookieDomin = ".metal.com";
exports.MetalCookieDomin = MetalCookieDomin;
var MMIPriceCookieDomin = ".mmiprices.com";
exports.MMIPriceCookieDomin = MMIPriceCookieDomin;

},{}],2:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
exports.__esModule = true;
var frame_1 = require("./frame");
var env_1 = require("../utils/env");
var uuid_1 = require("@smm/uuid");
var device_1 = require("./plugins/device");
var session_1 = require("./plugins/session");
var device_2 = require("./result_fn/device");
var time_1 = require("./result_fn/time");
var token_1 = require("./result_fn/token");
var session_2 = require("./result_fn/session");
var app_id_1 = require("./result_fn/app_id");
var path_1 = require("./result_fn/path");
var referrer_1 = require("./result_fn/referrer");
var domain_1 = require("./result_fn/domain");
var query_1 = require("./result_fn/query");
var scheme_1 = require("./result_fn/scheme");
var Sa = /** @class */ (function (_super) {
    __extends(Sa, _super);
    function Sa(isdev, plugins) {
        var _this = _super.call(this, isdev) || this;
        _this.events = [];
        for (var _i = 0, plugins_1 = plugins; _i < plugins_1.length; _i++) {
            var p = plugins_1[_i];
            p.Inject();
        }
        return _this;
    }
    Sa.prototype.setCommon = function (item) {
        this.common = item;
    };
    Sa.prototype.setFlow = function (item) {
        this.events.push(item);
    };
    Sa.prototype.setFlows = function (items) {
        this.events = items;
    };
    Sa.prototype.parse = function () {
        return this.parse_common(this.common, this.events);
    };
    Sa.prototype.parse_common = function (common, events) {
        var result = {};
        for (var key in common) {
            result[key] = common[key]();
        }
        //  events
        var flow = [];
        for (var i = 0; i < (events && events.length); i++) {
            var event_1 = events[i];
            var flow_item = {};
            for (var key in event_1) {
                flow_item[key] = event_1[key]();
            }
            flow.push(flow_item);
        }
        result["flow"] = flow;
        return result;
    };
    Sa.prototype.send_log_events = function (url, cb) {
        this.ajax({
            type: "POST",
            url: url ? url : this.event_url(),
            contentType: "application/json",
            dataType: "json",
            data: JSON.stringify(this.parse()),
            success: function () {
                cb && cb();
            }
        });
    };
    return Sa;
}(frame_1.Frame));
exports.Sa = Sa;
// set Default
var DefaultSa = new Sa(env_1.IsDev(), [
    device_1.DefaultDevicePlugin,
    session_1.DefaultSessionPlugin
]);
exports.DefaultSa = DefaultSa;
var common = {};
var commonKVArray = [
    token_1.tokenKV,
    session_2.sessionidKV,
    session_2.sessionstartidKV,
    time_1.timeKV,
    device_2.deviceKV,
    app_id_1.appIdKV
];
for (var _i = 0, commonKVArray_1 = commonKVArray; _i < commonKVArray_1.length; _i++) {
    var item = commonKVArray_1[_i];
    common[item.name] = item.value;
}
DefaultSa.setCommon(common);
var flow = {};
var flowEventKVArray = [
    {
        name: "flow_id",
        value: uuid_1.UUID.Generate
    },
    {
        name: "flow_type",
        value: function () { return "f"; }
    },
    {
        name: "timestamp",
        value: time_1.time
    },
    scheme_1.schemeKV,
    path_1.pathKV,
    query_1.queryKV,
    domain_1.domainKV,
    referrer_1.referrerKV
];
for (var _a = 0, flowEventKVArray_1 = flowEventKVArray; _a < flowEventKVArray_1.length; _a++) {
    var item = flowEventKVArray_1[_a];
    flow[item.name] = item.value;
}
DefaultSa.setFlow(flow);
if (!window.define) {
    window.smm_spider = Sa;
    window.smm_default_spider = function () {
        DefaultSa.send_log_events();
    };
}
var DefinedSa = function (flow_type, cb, flow_event_type_id, flow_data) {
    var sa = new Sa(env_1.IsDev(), [
        device_1.DefaultDevicePlugin,
        session_1.DefaultSessionPlugin
    ]);
    var common = {};
    var commonKVArray = [
        token_1.tokenKV,
        session_2.sessionidKV,
        session_2.sessionstartidKV,
        time_1.timeKV,
        device_2.deviceKV,
        app_id_1.appIdKV
    ];
    for (var _i = 0, commonKVArray_2 = commonKVArray; _i < commonKVArray_2.length; _i++) {
        var item = commonKVArray_2[_i];
        common[item.name] = item.value;
    }
    sa.setCommon(common);
    var flow = {};
    var flowEventKVArray = [
        {
            name: "flow_id",
            value: uuid_1.UUID.Generate
        },
        {
            name: "flow_type",
            value: function () { return flow_type; }
        },
        {
            name: "timestamp",
            value: time_1.time
        },
        scheme_1.schemeKV,
        path_1.pathKV,
        query_1.queryKV,
        domain_1.domainKV,
        referrer_1.referrerKV
    ];
    // e：事件类型
    if (/^e{1}$/i.test(flow_type)) {
        if (flow_event_type_id) {
            flowEventKVArray.push({
                name: "event_type_id",
                value: function () { return flow_event_type_id; }
            });
        }
        if (flow_data) {
            flowEventKVArray.push({
                name: "data",
                value: function () { return flow_data; }
            });
        }
    }
    for (var _a = 0, flowEventKVArray_2 = flowEventKVArray; _a < flowEventKVArray_2.length; _a++) {
        var item = flowEventKVArray_2[_a];
        flow[item.name] = item.value;
    }
    sa.setFlow(flow);
    sa.send_log_events(undefined, cb);
};
exports.DefinedSa = DefinedSa;
var EventType;
(function (EventType) {
    EventType["click"] = "click";
    EventType["mouseenter"] = "mouseenter";
    EventType["mouseover"] = "mouseover";
    EventType["mouseout"] = "mouseout";
    EventType["unblur"] = "unblur";
    EventType["change"] = "change";
    EventType["hover"] = "hover";
})(EventType || (EventType = {}));
exports.EventType = EventType;
var ClassDefinedSa = function (obj) {
    if (Object.prototype.toString.call(obj.className) === '[object Array]') {
        var list = obj.className;
        for (var i = 0; i < list.length; i++) {
            runDinedSa(obj);
        }
    }
    else {
        runDinedSa(obj);
    }
};
exports.ClassDefinedSa = ClassDefinedSa;
var runDinedSa = function (obj) {
    var className = obj.className;
    document.querySelector(className).addEventListener(obj.eventType, function () {
        DefinedSa(obj.flow_type, obj.cb, obj.event_type_id, obj.data);
    });
};
window.smm_defined_spider = DefinedSa;
window.ClassEventDefinedSa = ClassDefinedSa;

},{"../utils/env":18,"./frame":3,"./plugins/device":4,"./plugins/session":5,"./result_fn/app_id":6,"./result_fn/device":7,"./result_fn/domain":8,"./result_fn/path":10,"./result_fn/query":11,"./result_fn/referrer":12,"./result_fn/scheme":13,"./result_fn/session":14,"./result_fn/time":15,"./result_fn/token":16,"@smm/uuid":20}],3:[function(require,module,exports){
"use strict";
exports.__esModule = true;
var Frame = /** @class */ (function () {
    function Frame(isdev) {
        this.isdev = isdev;
    }
    Frame.prototype.host = function () {
        if (this.isdev) {
            return "https://testplatform.smm.cn";
        }
        return "https://platform.smm.cn";
    };
    //   事件统计
    Frame.prototype.event_url = function () {
        return this.host() + "/collectioncenter/log_event";
    };
    // 行为统计
    Frame.prototype.device_url = function () {
        return this.host() + "/collectioncenter/log_device";
    };
    // 封装原生ajax
    Frame.prototype.ajax = function (params) {
        var ajaxData = {
            type: params.type || "GET",
            url: params.url || "",
            async: params.async || "true",
            data: params.data || null,
            dataType: params.dataType || "text",
            contentType: params.contentType || "application/x-www-form-urlencoded",
            beforeSend: params.beforeSend || function () { },
            success: params.success || function () { },
            error: params.error || function () { }
        };
        ajaxData.beforeSend();
        var xhr = new XMLHttpRequest();
        xhr.responseType = ajaxData.dataType;
        xhr.open(ajaxData.type, ajaxData.url, ajaxData.async);
        xhr.setRequestHeader("Content-Type", ajaxData.contentType);
        xhr.send(this.convertData(ajaxData.data));
        xhr.onreadystatechange = function () {
            if (xhr.readyState == 4) {
                if (xhr.status == 200) {
                    ajaxData.success(xhr.response);
                }
                else {
                    ajaxData.error();
                }
            }
        };
    };
    Frame.prototype.convertData = function (data) {
        if (typeof data === 'object') {
            var convertResult = "";
            for (var c in data) {
                convertResult += c + "=" + data[c] + "&";
            }
            convertResult = convertResult.substring(0, convertResult.length - 1);
            return convertResult;
        }
        else {
            return data;
        }
    };
    return Frame;
}());
exports.Frame = Frame;

},{}],4:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
exports.__esModule = true;
var uuid_1 = require("@smm/uuid");
var cookie_1 = require("@smm/cookie");
var env_1 = require("../../utils/env");
var frame_1 = require("../frame");
var const_1 = require("../const");
var device_1 = require("../result_fn/device");
var height_1 = require("../result_fn/height");
var width_1 = require("../result_fn/width");
var Device = /** @class */ (function (_super) {
    __extends(Device, _super);
    function Device(isdev, resultkv) {
        var _this = _super.call(this, isdev) || this;
        _this.common = {};
        for (var _i = 0, resultkv_1 = resultkv; _i < resultkv_1.length; _i++) {
            var item = resultkv_1[_i];
            _this.common[item.name] = item.value;
        }
        return _this;
    }
    Device.prototype.log_device = function (url) {
        this.ajax({
            type: "POST",
            url: url ? url : this.device_url(),
            contentType: "application/json",
            dataType: "json",
            data: JSON.stringify(this.parse(this.common))
        });
    };
    Device.prototype.parse = function (common) {
        var result = {};
        for (var key in common) {
            result[key] = common[key]();
        }
        return result;
    };
    Device.prototype.Inject = function () {
        var id = cookie_1.Cookie.Get(const_1.DeviceCookieID);
        if (!id) {
            var smm_device_id = uuid_1.UUID.Generate();
            // 判断站点存放cookie
            if (env_1.IsMetal()) {
                // 英文站
                cookie_1.Cookie.Set(const_1.DeviceCookieID, smm_device_id, 365 * 10, const_1.MetalCookieDomin); // 设置10年
            }
            else if (env_1.IsMMIPrice()) {
                // MMIPrice
                cookie_1.Cookie.Set(const_1.DeviceCookieID, smm_device_id, 365 * 10, const_1.MMIPriceCookieDomin); // 设置10年
            }
            else {
                // 中文站 默认domain smm.cn
                cookie_1.Cookie.Set(const_1.DeviceCookieID, smm_device_id, 365 * 10); // 设置10年
            }
            this.log_device();
        }
    };
    return Device;
}(frame_1.Frame));
exports.Device = Device;
var DefaultDevicePlugin = new Device(env_1.IsDev(), [device_1.deviceKV, width_1.widthKV, height_1.heightKV]);
exports.DefaultDevicePlugin = DefaultDevicePlugin;

},{"../../utils/env":18,"../const":1,"../frame":3,"../result_fn/device":7,"../result_fn/height":9,"../result_fn/width":17,"@smm/cookie":19,"@smm/uuid":20}],5:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
exports.__esModule = true;
var uuid_1 = require("@smm/uuid");
var cookie_1 = require("@smm/cookie");
var env_1 = require("../../utils/env");
var frame_1 = require("../frame");
var const_1 = require("../const");
var Session = /** @class */ (function (_super) {
    __extends(Session, _super);
    function Session(isdev) {
        return _super.call(this, isdev) || this;
    }
    Session.prototype.Inject = function () {
        this._init();
        this._event();
    };
    // 初始化
    Session.prototype._init = function () {
        var session_id = cookie_1.Cookie.Get(const_1.SessionCookieID) || uuid_1.UUID.Generate();
        var now = new Date();
        var start_time = cookie_1.Cookie.Get(const_1.SessionStartTimeCookieID) || String(now.getTime());
        if (env_1.IsMetal()) {
            // 英文站
            this._cookieSet(session_id, start_time, const_1.MetalCookieDomin);
        }
        else if (env_1.IsMMIPrice()) {
            // MMIPrice
            this._cookieSet(session_id, start_time, const_1.MMIPriceCookieDomin);
        }
        else {
            // 中文站 默认domain smm.cn
            this._cookieSet(session_id, start_time);
        }
    };
    // 置放cookie
    Session.prototype._cookieSet = function (session_id, start_time, domain) {
        cookie_1.Cookie.Set(const_1.SessionCookieID, session_id, 1 / (24 * 2), domain ? domain : cookie_1.Cookie.DOMAIN); // 设置半小时
        cookie_1.Cookie.Set(const_1.SessionStartTimeCookieID, start_time, 1 / (24 * 2), domain ? domain : cookie_1.Cookie.DOMAIN); // 设置半小时
    };
    // 事件
    Session.prototype._event = function () {
        var _this = this;
        var filter = false;
        ['click', 'mouseover', 'keydown'].forEach(function (item) {
            document.addEventListener(item, function () {
                if (!filter) {
                    filter = true;
                    _this._init();
                    setTimeout(function () {
                        filter = false;
                    }, 1000);
                }
            });
        });
    };
    return Session;
}(frame_1.Frame));
exports.Session = Session;
var DefaultSessionPlugin = new Session(env_1.IsDev());
exports.DefaultSessionPlugin = DefaultSessionPlugin;

},{"../../utils/env":18,"../const":1,"../frame":3,"@smm/cookie":19,"@smm/uuid":20}],6:[function(require,module,exports){
"use strict";
exports.__esModule = true;
function AppID() {
    return 1;
}
exports.AppID = AppID;
var appIdKV = {
    name: "app_id",
    value: AppID
};
exports.appIdKV = appIdKV;

},{}],7:[function(require,module,exports){
"use strict";
exports.__esModule = true;
var cookie_1 = require("@smm/cookie");
var const_1 = require("../const");
function device() {
    return cookie_1.Cookie.Get(const_1.DeviceCookieID);
}
exports.device = device;
var deviceKV = {
    name: "device_id",
    value: device
};
exports.deviceKV = deviceKV;

},{"../const":1,"@smm/cookie":19}],8:[function(require,module,exports){
"use strict";
exports.__esModule = true;
function domain() {
    return window.location.host;
}
exports.domain = domain;
var domainKV = {
    name: "domain",
    value: domain
};
exports.domainKV = domainKV;

},{}],9:[function(require,module,exports){
"use strict";
exports.__esModule = true;
function height() {
    return window.innerHeight;
}
exports.height = height;
var heightKV = {
    name: "height",
    value: height
};
exports.heightKV = heightKV;

},{}],10:[function(require,module,exports){
"use strict";
exports.__esModule = true;
function path() {
    return window.location.pathname;
}
exports.path = path;
var pathKV = {
    name: "path",
    value: path
};
exports.pathKV = pathKV;

},{}],11:[function(require,module,exports){
"use strict";
exports.__esModule = true;
function query() {
    return window.location.search.substr(1) || "";
}
exports.query = query;
var queryKV = {
    name: "query",
    value: query
};
exports.queryKV = queryKV;

},{}],12:[function(require,module,exports){
"use strict";
exports.__esModule = true;
function referrer() {
    return document.referrer;
}
exports.referrer = referrer;
var referrerKV = {
    name: "referrer",
    value: referrer
};
exports.referrerKV = referrerKV;

},{}],13:[function(require,module,exports){
"use strict";
exports.__esModule = true;
function scheme() {
    return window.location.protocol;
}
exports.scheme = scheme;
var schemeKV = {
    name: "scheme",
    value: scheme
};
exports.schemeKV = schemeKV;

},{}],14:[function(require,module,exports){
"use strict";
exports.__esModule = true;
var cookie_1 = require("@smm/cookie");
var const_1 = require("../const");
function session_id() {
    return cookie_1.Cookie.Get(const_1.SessionCookieID);
}
exports.session_id = session_id;
function session_start_id() {
    return Number(cookie_1.Cookie.Get(const_1.SessionStartTimeCookieID));
}
exports.session_start_id = session_start_id;
var sessionidKV = {
    name: "session_id",
    value: session_id
};
exports.sessionidKV = sessionidKV;
var sessionstartidKV = {
    name: "session_start_timestamp",
    value: session_start_id
};
exports.sessionstartidKV = sessionstartidKV;

},{"../const":1,"@smm/cookie":19}],15:[function(require,module,exports){
"use strict";
exports.__esModule = true;
function time() {
    return new Date().getTime();
}
exports.time = time;
var timeKV = {
    name: "upload_timestamp",
    value: time
};
exports.timeKV = timeKV;

},{}],16:[function(require,module,exports){
"use strict";
exports.__esModule = true;
var cookie_1 = require("@smm/cookie");
var const_1 = require("../const");
function token() {
    return cookie_1.Cookie.Get(const_1.TokenCookieID);
}
exports.token = token;
var tokenKV = {
    name: "token",
    value: token
};
exports.tokenKV = tokenKV;

},{"../const":1,"@smm/cookie":19}],17:[function(require,module,exports){
"use strict";
exports.__esModule = true;
function width() {
    return window.innerWidth;
}
exports.width = width;
var widthKV = {
    name: "width",
    value: width
};
exports.widthKV = widthKV;

},{}],18:[function(require,module,exports){
"use strict";
exports.__esModule = true;
function IsDev() {
    // TODO 不建议使用
    return window.location.host.indexOf("test") !== -1;
}
exports.IsDev = IsDev;
function IsMetal() {
    return window.location.host.indexOf("metal.com") !== -1;
}
exports.IsMetal = IsMetal;
function IsMMIPrice() {
    return window.location.host.indexOf("mmiprices.com") !== -1;
}
exports.IsMMIPrice = IsMMIPrice;

},{}],19:[function(require,module,exports){
"use strict";
exports.__esModule = true;
var Cookie = /** @class */ (function () {
    function Cookie() {
    }
    Cookie.Set = function (key, value, expires, domain, path) {
        if (expires === void 0) { expires = 1; }
        if (!key) {
            return;
        }
        var cookie = key + "=" + encodeURIComponent(value);
        var date = new Date();
        date.setTime(date.getTime() + expires * 24 * 60 * 60 * 1000);
        cookie += ";expires=" + date.toUTCString();
        cookie += ";domain=" + (domain ? domain : Cookie.DOMAIN);
        cookie += ";path=" + (path ? path : Cookie.PATH);
        document.cookie = cookie;
    };
    Cookie.Get = function (key) {
        var reg = new RegExp("(^| )" + key + "=([^;]*)(;|$)");
        var arr = document.cookie.match(reg);
        if (arr) {
            return arr[2];
        }
        else {
            return null;
        }
    };
    Cookie.PATH = "/";
    Cookie.DOMAIN = ".smm.cn";
    Cookie.DefaultKey = "SMM_auth_token";
    return Cookie;
}());
exports.Cookie = Cookie;

},{}],20:[function(require,module,exports){
"use strict";
exports.__esModule = true;
var UUID = /** @class */ (function () {
    function UUID() {
    }
    UUID.Generate = function () {
        var s = [];
        for (var i = 0; i < 36; i++) {
            s[i] = UUID.HEX.substr(Math.floor(Math.random() * 0x10), 1);
        }
        s[14] = "5";
        s[19] = UUID.HEX.substr((s[19] & 0x3) | 0x8, 1);
        s[8] = s[13] = s[18] = s[23] = "-";
        return s.join("");
    };
    UUID.HEX = "0123456789abcdef";
    return UUID;
}());
exports.UUID = UUID;

},{}]},{},[2])
