// @smm/sa/dist/sa/frame.js
"use strict";
exports.__esModule = true;
var Frame = /** @class */ (function () {
    function Frame(isdev) {
        this.isdev = isdev;
    }
    Frame.prototype.host = function () {
        if (this.isdev) {
            return "https://testplatform.smm.cn";
        }
        return "https://platform.smm.cn";
    };
    //   事件统计
    Frame.prototype.event_url = function () {
        return this.host() + "/collectioncenter/log_event";
    };
    // 行为统计
    Frame.prototype.device_url = function () {
        return this.host() + "/collectioncenter/log_device";
    };
    // 封装原生ajax
    Frame.prototype.ajax = function (params) {
        var ajaxData = {
            type: params.type || "GET",
            url: params.url || "",
            async: params.async || "true",
            data: params.data || null,
            dataType: params.dataType || "text",
            contentType: params.contentType || "application/x-www-form-urlencoded",
            beforeSend: params.beforeSend || function () { },
            success: params.success || function () { },
            error: params.error || function () { }
        };
        ajaxData.beforeSend();
        var xhr = new XMLHttpRequest();
        xhr.responseType = ajaxData.dataType;
        xhr.open(ajaxData.type, ajaxData.url, ajaxData.async);
        xhr.setRequestHeader("Content-Type", ajaxData.contentType);
        xhr.send(this.convertData(ajaxData.data));
        xhr.onreadystatechange = function () {
            if (xhr.readyState == 4) {
                if (xhr.status == 200) {
                    ajaxData.success(xhr.response);
                }
                else {
                    ajaxData.error();
                }
            }
        };
    };
    Frame.prototype.convertData = function (data) {
        if (typeof data === 'object') {
            var convertResult = "";
            for (var c in data) {
                convertResult += c + "=" + data[c] + "&";
            }
            convertResult = convertResult.substring(0, convertResult.length - 1);
            return convertResult;
        }
        else {
            return data;
        }
    };
    return Frame;
}());
exports.Frame = Frame;
